File: src/main/java/com/github/t1/yaml/model/Mapping.java
package com.github.t1.yaml.model;

import lombok.Data;
import lombok.EqualsAndHashCode;

import java.util.ArrayList;
import java.util.List;

@EqualsAndHashCode(callSuper = true)
public @Data class Mapping extends Collection {
    private List<Entry> entries = new ArrayList<>();

    public Mapping entry(String key, String value) { return entry(new Scalar().line(key), value); }

    public Mapping entry(Scalar key, String value) { return entry(key, new Scalar().line(value)); }

    public Mapping entry(String key, Scalar value) { return entry(new Scalar().line(key), value); }

    public Mapping entry(String key, Node value) { return entry(new Scalar().line(key), value); }

    public Mapping entry(Node key, Node value) { return entry(new Entry().key(key).value(value)); }

    public Mapping entry(Entry entry) {
        entries.add(entry);
        return this;
    }

    @Override public void guide(Visitor visitor) {
        visitor.visit(this);
        for (Entry entry : entries) {
            visitor.enterMappingEntry(this, entry);
            entry.guide(visitor);
            visitor.leaveMappingEntry(this, entry);
        }
        visitor.leave(this);
    }

    public Entry lastEntry() { return entries.get(entries.size() - 1); }

    public static @Data class Entry {
        private boolean hasMarkedKey = false;
        private boolean hasNlAfterKey = false;
        private Node key;
        private Node value;

        void guide(Visitor visitor) {
            visitor.enterMappingKey(this, key);
            key.guide(visitor);
            visitor.leaveMappingKey(this, key);
            visitor.enterMappingValue(this, value);
            value.guide(visitor);
            visitor.leaveMappingValue(this, value);
        }
    }
}

File: src/main/java/com/github/t1/yaml/model/Mapping.kt
package com.github.t1.yaml.model

import com.github.t1.yaml.model.Collection.Companion.DEFAULT_STYLE
import com.github.t1.yaml.model.Collection.Style

data class Mapping(
    override var anchor: String? = null,
    override var spacing: String? = null,
    override var lineWrapping: String? = null,
    override var style: Style = DEFAULT_STYLE,
    val entries: MutableList<Entry> = mutableListOf()
) : Collection {

    fun entry(key: String, value: String): Mapping = entry(Scalar().line(key), value)

    fun entry(key: Scalar, value: String): Mapping = entry(key, Scalar().line(value))

    fun entry(key: String, value: Scalar): Mapping = entry(Scalar().line(key), value)

    fun entry(key: String, value: Node): Mapping = entry(Scalar().line(key), value)

    fun entry(key: Node, value: Node): Mapping = entry(Entry(key, value))

    fun entry(entry: Entry): Mapping {
        entries.add(entry)
        return this
    }

    val firstEntry get() = entries[0]
    val lastEntry get() = entries[entries.size - 1]

    override fun guide(visitor: Visitor) {
        visitor.visit(this)
        for (entry in entries) {
            visitor.enterMappingEntry(this, entry)
            entry.guide(visitor)
            visitor.leaveMappingEntry(this, entry)
        }
        visitor.leave(this)
    }

    data class Entry(
        var key: Node? = null,
        var value: Node? = null,
        var hasMarkedKey: Boolean = false,
        var hasNlAfterKey: Boolean = false
    ) {
        fun guide(visitor: Visitor) {
            visitor.enterMappingKey(this, key!!)
            key!!.guide(visitor)
            visitor.leaveMappingKey(this, key!!)
            visitor.enterMappingValue(this, value!!)
            value!!.guide(visitor)
            visitor.leaveMappingValue(this, value!!)
        }
    }
}
